pkg/fqdn/rulegen.go

```diff
@@ -266,83 +180,65 @@ func (gen *RuleGen) GetDNSNames() (dnsNames []string) {
 
 // UpdateGenerateDNS inserts the new DNS information into the cache. If the IPs
 // have changed for a name, store which rules must be updated in rulesToUpdate,
-// regenerate them, and emit via AddGeneratedRulesAndUpdateSelectors.
+// regenerate them, and emit via UpdateSelectors.
 func (gen *RuleGen) UpdateGenerateDNS(lookupTime time.Time, updatedDNSIPs map[string]*DNSIPRecords) error {
 	// Update IPs in gen
-	uuidsToUpdate, updatedDNSNames := gen.UpdateDNSIPs(lookupTime, updatedDNSIPs)
+	fqdnSelectorsToUpdate, updatedDNSNames := gen.UpdateDNSIPs(lookupTime, updatedDNSIPs)
 	for dnsName, IPs := range updatedDNSNames {
 		log.WithFields(logrus.Fields{
-			"matchName":     dnsName,
-			"IPs":           IPs,
-			"uuidsToUpdate": uuidsToUpdate,
+			"matchName":             dnsName,
+			"IPs":                   IPs,
+			"fqdnSelectorsToUpdate": fqdnSelectorsToUpdate,
 		}).Debug("Updated FQDN with new IPs")
 	}
 
-	// Generate a new rule for each sourceRule that needs an update.
-	rulesToUpdate, notFoundUUIDs := gen.GetRulesByUUID(uuidsToUpdate)
-	if len(notFoundUUIDs) != 0 {
-		log.WithField("uuid", strings.Join(notFoundUUIDs, ",")).
-			Debug("Did not find all rules during update")
-	}
-	generatedRules, namesMissingIPs, selectorIPMapping := gen.GenerateRulesFromSources(rulesToUpdate)
+	namesMissingIPs, selectorIPMapping := gen.GenerateSelectorUpdates(fqdnSelectorsToUpdate)
 	if len(namesMissingIPs) != 0 {
 		log.WithField(logfields.DNSName, namesMissingIPs).
 			Debug("No IPs to insert when generating DNS name selected by ToFQDN rule")
 	}
 
-	// TODO (ianvernon) if no new rules to add, add additional logic in function
-	// below?
-
-	// emit the new rules
-	return gen.config.AddGeneratedRulesAndUpdateSelectors(generatedRules, selectorIPMapping, namesMissingIPs)
+	return gen.config.UpdateSelectors(selectorIPMapping, namesMissingIPs)
 }
 
 // ForceGenerateDNS unconditionally regenerates all rules that refer to DNS
 // names in namesToRegen. These names are FQDNs and toFQDNs.matchPatterns or
 // matchNames that match them will cause these rules to regenerate.
 func (gen *RuleGen) ForceGenerateDNS(namesToRegen []string) error {
-	affectedRulesSet := make(map[string]struct{}, 0)
+	// Lock needed...?
```

Comment:
With the older code, no, since `gen.sourceRules` had it's own locking. The new setup might be a little annoying. `gen.allSelectors` does seem to need a lock... but GenerateSelectorUpdates also takes the lock :/ It (`gen.GenerateSelectorUpdates`) is only called in contexts that could/need to hold the lock, so it might be simplest to pop taking the lock one level up the stack.

Reply:
Will do.

Reply:
Done!